posts/2023-04-13-rust-by-examples.adoc

```diff
@@ -0,0 +1,726 @@
+}
+
+fn test_animals() {
+    let chicken = Chicken("Jenny".to_string());
+    let fox = Fox("Joe".to_string());
+
+    assert_animal(&chicken, "Jenny", "chicken");
+    assert_animal(&fox, "Joe", "Fox");
+}
+----
+
+A bit more complex example, reusing TitleBasics struct from the previous section:
+[source,rust]
+----
+// the following trait abstracts the access to a database row.
+// It may be implemented for different databases or for unit testing without a database.
+pub trait DbRow {
+    fn opt_string(&self, column: &str) -> Option<String>;
+    fn opt_i32(&self, column: &str) -> Option<i32>;
+}
+
+// add a from_db_row method to TitleBasics
+impl TitleBasics {
+    pub fn from_db_row(r: &dyn DbRow) -> TitleBasics {
+        TitleBasics {
+            id: r.opt_string("tconst").unwrap(),
+            title_type: r.opt_string("titletype"),
+            primary_title: r.opt_string("primarytitle"),
+            start_year: r.opt_i32("startyear"),
+        }
+    }
+}
+
+// Implement DbRow for Postgres
+// Although PgRow is coming from an external library, we can extend it,
+// a bit like implicit classes in Scala2
+use rocket_db_pools::sqlx::{Row, postgres::PgRow};
+impl DbRow for PgRow {
+    fn opt_string(&self, column: &str) -> Option<String> {
+        self.try_get::<String, &str>(column).ok()
+    }
+    fn opt_i32(&self, column: &str) -> Option<i32> {
+        self.try_get::<i32, &str>(column).ok()
+    }
+}
+
+// Use DB row for querying a DB table, with the rocket_db_pools library
+// "async" is an asynchronous function, practically meaning it returns a Future
+use rocket_db_pools::sqlx;
+use rocket_db_pools::sqlx::{Error, PgPool, postgres::PgRow};
+pub async fn query_title_basics(db_pool: &PgPool, id: &str) -> Result<TitleBasics, Error> {
+    sqlx::query("SELECT * FROM title_basics WHERE tconst = $1")
+        .bind(id)
+        .fetch_one(db_pool)
+        .await
+        .and_then(|row: PgRow| Ok(TitleBasics::from_db_row(&row)))
+}
+
+
+// A mock DB row used for unit testing
+struct TestDbRow<'r> {
+    map: HashMap<&'static str, &'r str>,
+}
+
+// Implement DbRow for the mock DB row
+impl<'r> DbRow for TestDbRow<'r> {
+    fn opt_string(&self, column: &str) -> Option<String> {
+        self.map.get(column).map(|x| x.to_string())
+    }
+    fn opt_i32(&self, column: &str) -> Option<i32> {
+        self.map.get(column).map(|x| x.parse::<i32>().unwrap())
+    }
+}
+
+// Test TitleBasics::from_db_row() without a database:
+// #[test] is an annotation macro
+#[test]
+fn test_title_basics_from_db_row() {
+    let values = HashMap::from([("tconst", "abcd")]);
+    let test_row = TestDbRow { map: values };
+    let title_basics = TitleBasics::from_db_row(&test_row);
+
+    assert!(title_basics.id == "abcd");
+    // ...
+}
+----
+
+Another, less-useful polymorphism is operator overloading. You can find the list of overloadable operators https://doc.rust-lang.org/book/appendix-02-operators.html[here^]
```

Comment:
Instead of less-useful would you consider rephrasing it as: "Another less common polymorphism is operator overload".
"Less-useful" suggest is a lesser alternative than bounded polymorphism